src/iam-token.js

```diff
@@ -0,0 +1,142 @@
+        apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    }
+    if (!apiKey) {
+      return Promise.reject('Failed to find valid apikey to authenticate against iam');
+    }
+
+    let res = await this._requestToken(iam, apiKey);
+    return res.access_token;
+  }
+
+
+  async _requestToken(iam, apiKey) {
+    let token;
+
+    const apiKeyHash = hash(apiKey, { algorithm: 'shake256' });
+      if (token === undefined && objectPath.has(this.s3TokenCache, [apiKeyHash])) {
+      // fetch cached token
+        token = objectPath.get(this.s3TokenCache, [apiKeyHash]);
+        this.log.info(`MASCD got the cached token`);
+    }
+    if (token !== undefined) {
+      const expires = objectPath.get(token, 'expiration', 0); // expiration: time since epoch in seconds
+      // re-use cached token as long as we are more than 2 minutes away from expiring.
+      if (Date.now() < (expires - 120) * 1000) {
+        this.log.info(`MASCD cached token has not expired, returning it`);
+        return token;
+      } else {
+        this.log.info(`MASCD cached token expired getting new one`);
+      }
+    }
+
+    if (waitingForToken === true) {
+      this.log.info(`MASCD waiting for different event to get a token`);
+      while (waitingForToken === true) {
+        this.log.info(`MASCD going to sleep`);
+        await sleep(2000);
+        this.log.info(`MASCD done with sleep`);
+      }
+      token = objectPath.get(this.s3TokenCache, [apiKeyHash]);
+      const expires = objectPath.get(token, 'expiration', 0);
+      this.log.info(`MASCD token expires ${expires}`)
+      if (Date.now() < (expires - 120) * 1000) {
+          this.log.info(`MASCD cached token got by other event has not expired, returning it`);
+          return token;
+      } else {
+          this.log.info(`MASCD cached token got by other event expired getting new one`);
+      }
+
+    } else {
+        this.log.info(`MASCD going to get the new token, setting waiting for token to true`)
+        waitingForToken = true
+    }
+
+    try {
+      let res = await axios({
+        method: 'post',
+        url: iam.url, // 'https://iam.cloud.ibm.com/identity/token',
+        params: {
+          'grant_type': iam.grantType || iam.grant_type, // 'urn:ibm:params:oauth:grant-type:apikey',
+          'apikey': apiKey
+        },
+        // data: `grant_type=urn:ibm:params:oauth:grant-type:apikey&apikey=${apiKey}`,
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded'
+        },
+        timeout: 60000
+      });
+      token = res.data;
+      try {
+        objectPath.set(this.s3TokenCache, [apiKeyHash], token);
+      } catch (fe) {
+        waitingForToken = false;
+        return Promise.reject(`failed to cache s3Token to disk at path ${tokenCacheFile}`, fe);
+      }
+      waitingForToken = false;
+      this.log.info(`MASCD returning new token`);
```

Comment:
Since this is a public razee repo, MASCD doesn't mean anything in this context. A more generic log message should be used. This goes for all the log messages in this file.